server/src/main/java/org/jboss/as/server/BootstrapListener.java

```diff
@@ -68,27 +68,34 @@ public StabilityMonitor getStabilityMonitor() {
         return monitor;
     }
 
-    public void printBootStatistics(String message) {
+    public StabilityStatistics getBootStatistics() {
         final StabilityStatistics statistics = new StabilityStatistics();
         try {
             monitor.awaitStability(statistics);
         } catch (final InterruptedException e) {
             Thread.currentThread().interrupt();
-            return;
+            return null;
         } finally {
             serviceTarget.removeMonitor(monitor);
-            final long bootstrapTime = System.currentTimeMillis() - startTime;
-            done(bootstrapTime, statistics, message);
+            futureContainer.done(serviceContainer);
             monitor.clear();
+            return statistics;
+        }
+    }
+
+    public void printBootStatisticsMessage(StabilityStatistics statistics, String message) {
+        if (statistics == null) {
+            return;
         }
+        final long bootstrapTime = System.currentTimeMillis() - startTime;
+        done(bootstrapTime, statistics, message);
```

Comment:
This has an unwanted side effect since you will be invoking `createStartupMarker` in the `postBoot()` phase. That will occur after the process has transitioned to `RUNNING`.

You have split the methods into two, but you have to just calculate the String message in the first method, store it, and print it in later the second method, which will be invoked on the `postBoot()`. 

All should behave as it was before, it is just you will be calculating the final message first to be printed later. We do not want to introduce any side effects, basically, changing when the message is printed is not enough to support any behavioral change.

Reply:
My apologies. I see what you mean. Thank you.
I force-pushed another change. I didn't want to store the whole message in the BootstrapListener instead of Logger, so I introduced a bit different approach. What do you think about that? Can it stay like that, or the potential of changing the statistics variable is just a no-go?

Reply:
Hi @khermano , you have progressed with it since now we do not have unexpected side effects that change how they currently behave. However, I was thinking of a slightly different approach I think it is more simple since doesn't expose data outside of the `boostrapListener`:

You can use the `ServerLogger.AS_ROOT_LOGGER.startedClean()` and `ServerLogger.AS_ROOT_LOGGER.startedWitErrors()` to just generate the messages, which could be Strings. You have already done this in the past, just create the loggers without message id and without a log level.

So in the first phase, you will prepare these messages that `printBootStatisticsMessage` will print later invoked from the `postBoot()`.

For example, you can create two different fields for the messages in `boostrapListener`. If you use two strings, then you will know later if the `printBootStatisticsMessage` is dealing with an ERROR or an INFO message. For example, you can create these two loggers by copying the existing ones (these loggers will not use a Message ID and log level):

`ServerLogger.AS_ROOT_LOGGER.prepareStartedClean(prettyVersion, bootstrapTime, started, active + passive + onDemand + never + lazy, onDemand + passive + lazy, message)` -> Will generate the info message, you can store it in a class field variable, e.g  String startedCleanMessage;

`ServerLogger.AS_ROOT_LOGGER.startedWitErrors(prettyVersion, bootstrapTime, started, active + passive + onDemand + never + lazy, failed + problem, onDemand + passive + lazy, message)` -> Will generate the Error message, you can store it in a class field variable, e.g  String startedWitErrorsMessage;

Then in `printBootStatisticsMessage()` print the one that is available and also reset them, e.g:
```
        if (!startedCleanMessage != null) {
            ServerLogger.AS_ROOT_LOGGER.startedClean(startedCleanMessage);
        } else if (!startedWitErrorsMessage.isEmpty()) {
            ServerLogger.AS_ROOT_LOGGER.startedWitErrors(startedWitErrorsMessage);
        }
       startedCleanMessage = null;
       startedWitErrorsMessage = null;
```

You will need to adapt the existing loggers `ServerLogger.AS_ROOT_LOGGER.startedClean(...)` and `ServerLogger.AS_ROOT_LOGGER.startedWitErrors(...)` so they use a single string as a message (the one generated by the prepareXXXX methods described before). These loggers will be the ones that will have a Message ID and a Log level message.

Reply:
Hi @yersan 
Thank you for the information and your suggestions. I force-pushed the changes. :)